### ISSUE TITLE
More flexibility in quantile mapping

### ISSUE THREAD

**Issue by DamienIrving** [Ouranosinc/xclim#1374]
### Addressing a Problem?

I wanted to flag a potential need for more flexibility in the way the quantile mapping functionality is implemented in xclim.

At the moment, after you run the training step...
```
QM = sdba.EmpiricalQuantileMapping.train(ref, hist)
``` 
... you end up with an object containing adjustment factors (`QM.ds['af']`) and historical quantiles (`QM.ds['hist_q']`). If you then go ahead and run `QM.adjust(sim)`, it uses `af` and `hist_q` to apply the quantile adjustments to the `sim` data.

In the context of bias correction, `ref` represents observational data and `sim` the model data to be corrected, whereas in the context of using a quantile delta change method to produce projections data `ref` is model data from a future model experiment and `sim` is the observational data that model delta changes are to be applied to.

Especially in the context of quantile delta change approaches (e.g. [Anandh et al 2011](https://doi.org/10.1029/2010WR009104), [Willems & Vrac, 2011](https://doi.org/10.1016/j.jhydrol.2011.02.030)) but also sometimes for bias correction, `sim_q` is actually what is needed as opposed to `hist_q`. In other words, adjustment factors are selected with reference to the quantiles of the data to be adjusted as opposed to the historical data (which is nice because it means no extrapolation is required).  

### Potential Solution

To get around the problem described above, I've been calculating `sim_q` using `sdba.nbutils.quantile` and then I replace the values in `QM.ds['hist_q']` with my `sim_q` data before running `QM.adjust(sim)`. This isn't a particularly elegant solution because I don't think `sdba.nbutils.quantile` is really meant to be used by general xclim users.

I'm wondering whether it would be better if `QM.ds` didn't contain `hist_q` and instead there was a `quantiles` argument for the `adjust` method so that either approach could be adopted? e.g.
```
sim_q = sdba.quantile(sim)
QM.adjust(sim, quantiles=sim_q)
```
or
```
hist_q = sdba.quantile(hist)
QM.adjust(sim, quantiles=hist_q)
```
(I'm envisaging a user-friendly quantiles function, perhaps `sdba.quantile`)

### Additional context

I'd be happy to help with a pull request to implement these suggested changes but I'd probably require some assistance (I've found it a little hard to follow exactly how it's all implemented internally).

### Contribution

- [X] I would be willing/able to open a Pull Request to contribute this feature.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by aulemahal** [comment]
What you are suggesting with `sim_q`  looks a lot like Quantile Delta Mapping, no ?

In EQM, `hist_q` is used to go from values of `sim` to quantiles which are then used to get the relevant factor (`af`). The match between values of  `sim` and `hist_q` is done through interpolation (nearest, linear, etc).

In QDM, there's no use for `hist_q` as the quantile of each value of `sim` is computed and then used to get the relevant factor.

AFAIU, the difference with what you are suggesting is that here the "matching" step is here done between the sim quantile and the `af` quantile coordinate. Whereas your suggestion uses the method of EQM, we match the value of sim with `sim_q` to get a quantile. I wouldn't expect the difference to be really significant ? I might be wrong!

**Comment by coxipi** [comment]
Hi @DamienIrving , thanks for the suggestion. I also agree that the idea of using `sim_q`  seems similar to QDM. 

Looking specifically at sec.3.2 of [Anandh et al 2011](https://doi.org/10.1029/2010WR009104), I don't think this is quite like QDM though. 

The first way working mentioned above is to correct the model 1):
```
QM = sdba.EmpiricalQuantileMapping.train(ref, hist)
sim_corrected = QM.adjust(sim)
```
Now, reading [Anandh et al 2011](https://doi.org/10.1029/2010WR009104), I understand we could simply swap `ref` and `sim` altogether to project the observations in the future 2):
```
QM = sdba.EmpiricalQuantileMapping.train(sim, hist)
ref_future = QM.adjust(ref)
```
So in both cases, I would say that indeed `hist_q` is needed. You interpolate the couple (`hist_q`, `af_q`) to obtain a function `AF(x)`.  This is the function found in the adjustment.

1. The training is for the transfer `hist` -> `ref`  (correct the model).  AF is used to obtain `sim` -> `sim_corrected`.
2. The training is for the transfer `hist` -> `sim` (project historical period in future).   AF is used to obtain `ref` -> `ref_future`.

Did I misunderstand the delta change I pointed at in the first paper? Or do you have another method in mind that does not correspond to the second method I've listed above (something more similar to QDM as mentioned by @aulemahal )?

**Comment by DamienIrving** [comment]
Thanks, @aulemahal and @coxipi.

Elements of both your responses are consistent with what I was trying to describe. I think my explanation was a little unclear as I was switching the meaning of `ref` and `sim` depending on whether the application was bias correction or delta change. Let me try again...

### Scenario 1: Bias correction
In this scenario you want to bias correct some model data (`sim`).
```
QM = sdba.EmpiricalQuantileMapping.train(ref, hist)
sim_corrected = QM.adjust(sim)
```
As it stands (as @aulemahal describes), in this situation `hist_q` is used to go from values of `sim` to quantiles which are then used to get the relevant factor (`af`). The use of `hist_q` here is consistent with how some authors perform bias correction, but an alternative (as @aulemahal notes) is to not use `hist_q` and instead compute the quantile of each value of `sim` (i.e. `sim_q`) and then use it to get the relevant factor. As @coxipi notes, I don't think that is exactly QDM.

### Scenario 2: Delta change
In this scenario you want to apply delta changes (from a model) to observations (`ref`):
```
QM = sdba.EmpiricalQuantileMapping.train(sim, hist)
ref_future = QM.adjust(ref)
```
As described in papers like [Anandh et al 2011](https://doi.org/10.1029/2010WR009104), this approach typically does not involve using `hist_q`. Instead, the quantile of each value of `ref` is computed (i.e. `ref_q`) and then used to get the relevant factor.

One advantage of not using `hist_q` (in either scenario) is that no extrapolation is needed.

**Comment by coxipi** [comment]
@DamienIrving Yes, that's clear. I didn't read the paper properly. As I understand now, the CF (here `af`) factors found in the training are to be used as-is for the adjustment part. You can obtain this with `QuantileDeltaMapping`:
```
QDM = sdba.QuantileDeltaMapping.train(sim, hist)
ref_future = QDM.adjust(ref)
```

## Comparing QM & QDM
Let me compare QM and QDM by adopting the delta change terminology here (train `hist` on `sim`,  adjust `ref`):
* QM interpolates on quantile couples `(hist_q, af_q)` to find a function `AFQM`. Adjustment factors  appropriate for the ranked values `ref_n` of `ref` are found   with `af_n = AFQM(ref_n)` (ranks are like quantiles, but there are as many ranks as values in a dataset).
* QDM interpolates  `(q, af_q)` to find a function AFQDM.  In this case, the function is applied on ranks `n` related to ranked values `ref_n`  ,  `af_n = AFQDM(n)`. As a result, the adjustment factors are the same as those found in the training, as suggested in the paper. This is what you described I believe:
> Instead, the quantile of each value of ref is computed (i.e. ref_q) and used to get the relevant factor.

## Comparing the implementations (XClim vs. Damien's suggestion)
To bounce back on @aulemahal , the idea is basically the same. There might be a small difference between both implementations. In XClim, QDM trains `(q, af_q)` to get:

$$ \text{af}_{n}=\text{AF}(n) , \quad \text{XClim implementation}$$

What is suggested here is to train  `(ref_q, af_q)` to get:

$$ \text{af}'_{n}=\text{AF'}(ref_n), \quad \text{Damien's suggestion} $$

Since the ranked value $ref_n$ can be obtained from $n$ with an inverse CDF transform $ref_n=F^{-1}(n)$, this last equation can be written as  

$$ \text{af}'_{n}=\text{AF'}\circ F^{-1} (n) $$

The adjustement factors are the same if

$$ \text{AF} =  \text{AF'}\circ F^{-1} $$

Both procedures could lead to subtle differences, but in essence accomplish the same idea.

**Comment by coxipi** [comment]
Just to complete the idea that both implementations are not the same, this is basically:

Damien's

$$
(x_{i},y_{i}) \to_{interp} \quad y=f(x)
$$

vs. XClim's

$$
(g(x_{i}),y_{i})	\to_{interp}  \quad  y'=f'(g(x))=(f'\circ g)(x)
$$

If $g$ is linear in $x$, I think those are the same. Otherwise, no.

Using `y_i = x_i = {0,1,2,3,...9}`
![image](https://github.com/Ouranosinc/xclim/assets/71575674/cc3ace6a-0805-4fdf-a717-1d1cbd8faa9b)

Linear `g` gives the same answer, but not the non-linear one
![image](https://github.com/Ouranosinc/xclim/assets/71575674/44076ade-e249-4659-b6ae-eb377a1dae00)

**Comment by aulemahal** [comment]
Well, yes I agree with @coxipi that the two are not "the same".

However, from a practical view point, I'm not sure we care about the difference that much ?

Not so artificial example:
Using names as in this code:
```
FF = train(ref, hist)
sim_adj = FF.adjust(sim)
```
where all variables are temperatures.

After training:
```
q = [0, 0.01, 0.02, ...] # up to 1
hist_q = [-20, -15, -13,  ...] # not necessary
af = [1, 2, 3,  ...]
...
sim_q = [-18, -13, -11,  ...]
```

Method 1, QM as the suggested implementation: For value  -14 in sim, we see that the nearest value in `sim_q` is -13, which correspons to quantile 0.01, thus the factor to use is 2.

Method 2, QDM. For value -14 in sim, we get a rank that corresponds to a quantile of 0.05. Nearest quantile if 0.01, thus the factor to use is 2.

In my example, numbers fall neatly and I used a "nearest" interpolator. Of course, nuances are added in real world example and especially if using a linear or polynomial interpolator.  And indeed, that's what @coxipi  showed with his synthetic examples of `g`. But I still think QDM is close enough.

**Comment by coxipi** [comment]
> that's what @coxipi showed with his synthetic examples of g
Sorry, I removed the post with synthetic $g$ since I thought there was maybe an error, but anyways, I think your explanation here is clear.

**Comment by DamienIrving** [comment]
Thanks, @coxipi and @aulemahal. This is very helpful - I think I had been misunderstanding the definition of QDM in [Cannon et at 2015](https://journals.ametsoc.org/view/journals/clim/28/17/jcli-d-14-00754.1.xml).

From your description it sounds like the hack I've been implementing (described above), which is essentially:
```
QM = sdba.EmpiricalQuantileMapping.train(sim, hist)
QM.ds['hist_q'] = sdba.nbutils.quantile(ref)  #hack to replace hist_q with ref_q
ref_future = QM.adjust(ref)
```
is more or less (although technically not exactly) the same as just running:
```
QDM = sdba.QuantileDeltaMapping.train(sim, hist)
ref_future = QDM.adjust(ref)
```

If my understanding is correct, the `extrapolation` argument isn't really needed in these cases? In the QM case you can't have a value of `ref` beyond the bounds of `ref_q` so no extrapolation would be needed, and in the QDM case the methods works with the ranks as opposed to rank values so there would also be no situations where extrapolation would be required?

**Comment by DamienIrving** [comment]
I guess no changes need to be made to the xclim code as I has initially suggested, but perhaps I should submit a pull request to the xclim documentation to explain a bit more about `QuantileDeltaMapping`?

In particular, depending on the context it appears there are few different names for the method used by `QuantileDeltaMapping`:
- If quantile delta changes (e.g. between an historical and future model simulation) are applied (either additively or multiplicatively) to observational data in order to produce climate projection data for a future time period, the method has been referred to as *Quantile Delta Mapping* (QDM; [Cannon et al 2015](https://doi.org/10.1175/JCLI-D-14-00754.1)).
- If quantile biases (e.g. between an historical model simulation and observations) are removed from model data in order to produce bias corrected model data, the method has been referred to as *equidistant CDF matching* (EDCDFm; in the case of additive bias correction; [Li et al, 2010](https://doi.org/10.1029/2009JD012882)) or *equiratio CDF matching* (EQCDFm; in the case of multiplicative bias correction; [Wang and Chen, 2013](https://doi.org/10.1002/asl2.454)).

As explained by [Cannon et al (2015)](https://doi.org/10.1175/JCLI-D-14-00754.1), additive/multiplicative Quantile Delta Mapping is actually equivalent to equidistant/equiratio CDF matching. In other words, you get the same result regardless of whether you apply model delta changes to observational data or bias correct model data.

This means that 
```python
QDMadd = sdba.QuantileDeltaMapping.train(sim, hist, kind='+')
future_data = QDMadd.adjust(ref)
```
is equivalent to
```python
EDCDFm = sdba.QuantileDeltaMapping.train(ref, hist, kind='+')
future_data = EDCDFm.adjust(sim)
```

and 
```python
QDMmul = sdba.QuantileDeltaMapping.train(sim, hist, kind='*')
future_data = QDMmul.adjust(ref)
```
is equivalent to
```python
EQCDFm = sdba.QuantileDeltaMapping.train(ref, hist, kind='*')
future_data = EQCDFm.adjust(sim)
```

**Comment by aulemahal** [comment]
Hi @DamienIrving, we were thinking of closing this issue, but if you indeed have time to add some documentation as you suggested, that would be greatly appreciated!

**Comment by DamienIrving** [comment]
Yes, I've got some detailed notes on quantile delta mapping (QDM) and equidistant/ratio CDF matching (which can both be achieved using `sdba.QuantileDeltaMapping`)... I'd just need to take a look at where they could be adapted/inserted into the xclim docs.  
https://github.com/climate-innovation-hub/qqscale/blob/master/docs/method_ecdfm.md  
https://github.com/climate-innovation-hub/qqscale/blob/master/docs/method_qdm.md

**Comment by isaachcamp** [comment]
Hi all, apologies for bringing this back up again but I just wanted to clarify something. 

If I'm understanding the thread above correctly (and the qdm_adjust function), the xclim implementation of QDM interpolates the adjustment factors onto the future quantiles (equation (5) of [Cannon et al 2015](https://doi.org/10.1175/JCLI-D-14-00754.1)), but it does not apply the multiplicative factor between model historical and future quantiles, as in equation (6) of [Cannon et al 2015](https://doi.org/10.1175/JCLI-D-14-00754.1)?
